nginx/modsecurity/ngx_http_modsecurity.c

```diff
@@ -819,57 +761,72 @@ ngx_http_modsecurity_prevention_task_offload(ngx_http_request_t *r)
     if (ngx_http_modsecurity_load_request(r) != NGX_OK ||
             ngx_http_modsecurity_load_headers_in(r) != NGX_OK) {
         return NGX_ERROR;
     }
 
-    if (modsecIsRequestBodyAccessEnabled(req) && (r->headers_in.content_length || r->headers_in.chunked)) {
+    // ModSecurity only initializes the internal request context upon the call
+    // to modsecProcessRequestHeaders().
+    // In our case, we postpone this call so we need to initialize the context
+    // explicitly because predicates like modsecIsRequestBodyAccessEnabled()
+    // use it for the checks they perform.
+    const char *err = modsecInitializeRequestContext(ctx->req);
+    if (err) {
+        ngx_log_error(NGX_LOG_ERR, r->connection->log, 0, err);
+        return NGX_ERROR;
+    }
+
+    if (modsecIsRequestBodyAccessEnabled(ctx->req) && (r->headers_in.content_length || r->headers_in.chunked)) {
         if (ngx_http_modsecurity_load_request_body(r) != NGX_OK) {
             return NGX_ERROR;
         }
     }
 
-    ngx_thread_task_t *task = apr_thread_task_alloc(req->pool, sizeof(ngx_http_modsecurity_detection_thread_ctx_t));
+    ngx_thread_task_t *task = ngx_thread_task_alloc(r->pool, sizeof(ngx_http_modsecurity_thread_ctx_t));
     if (task == NULL) {
         return NGX_ERROR;
     }
 
-    ngx_http_modsecurity_detection_thread_ctx_t *thread_ctx = task->ctx;
-    thread_ctx->connection = ctx->connection;
-    thread_ctx->req = ctx->req;
-    thread_ctx->brigade = ctx->brigade;
+    ngx_http_modsecurity_thread_ctx_t *thread_ctx = task->ctx;
+    thread_ctx->r = r;
 
     task->handler = ngx_http_modsecurity_detection_thread_func;
     task->event.handler = ngx_http_modsecurity_detection_thread_completion;
-    task->event.data = thread_ctx;
+    task->event.data = r;
 
     ngx_thread_pool_t* thread_pool = ngx_thread_pool_get((ngx_cycle_t *)ngx_cycle, &thread_pool_name);
     if (ngx_thread_task_post(thread_pool, task) != NGX_OK) {
         return NGX_ERROR;
     }
 
-    // We need to reset these so they don't get cleaned up when the request is complete.
-    // This is important as the background task can outlive the request and we need to
-    // make sure the memory it uses is still alive.
-    ctx->connection = NULL;
-    ctx->req = NULL;
-    ctx->brigade = NULL;
-
-    return NGX_OK;
+    /* Increment request reference count to make sure it lives long enough
+     * for the task to be completed.
+     * Note that it is safe to do that after posting the task to the thread pool because
+     * we decrement it in the completion handler and it will only run after the current code
+     * as it is scheduled on the same (main) thread.
+     * Besides the reference counter ('count') we need to increment 'blocked' because this
+     * guarantees that the request is not terminated when the connection is closed -
+     * see ngx_http_terminate_handler() for how it neglects the 'count' value in this case.
+     */
+    ++r->main->count;
```

Comment:
in the completion, we only decrease block, so where the count got decremented. 
and i think count and block are used to control the lifecycle  of different struct, count is for the ctx and the block is for the request, correct me if i am wrong, so right now both ctx and request will be released at the same time.

Reply:
The count is decremented inside the call to `ngx_http_finalize_request()`. This is the way advertized by Nginx development guide: https://nginx.org/en/docs/dev/development_guide.html#http_request (see the "count" field).

Both `count` and `blocked` belong to the `ngx_http_request_t` structure. All allocations are done from the pool associated with an HTTP request so it is sufficient to guarantee that the request lives long enough because in this case its pool is not destroyed too early.